[EOC-496] Change filter and sort options to search

adamklepacz opened the pull request:
It is now possible to serach using whatever field we want (name, authorName) etc.


![chrome-capture (1)](https://user-images.githubusercontent.com/27632432/65684079-0ca27880-e05f-11e9-9e23-17db7ab5ce4a.gif)

alemikolo commented on client/src/modules/list/components/ItemsContainer.jsx:
```diff
@@ -1,94 +1,48 @@
-  AUTHOR: 'author',
-  VOTES: 'votes'
-});
+// FIXME: Remove sortbox and filterBox components and styles files
```
remove comment

alemikolo commented on client/src/modules/list/components/ItemsContainer.jsx:
```diff
@@ -125,18 +78,15 @@ class ItemsContainer extends Component {
+            <Filter
+              buttonContent={<CloseIcon />}
+              classes="items__filter"
+              fields={filterByFields}
```
hmm, `fields` prop is good because it is short but now I wonder if it is descriptive... Maybe sth like `fieldsToMatch`, `filterByFields`?

adamklepacz commented:
> I think `search input` is not enough different from the `add item input` or it is too close to it. Maybe a little bigger gaps between `search input`, `add input` and list title ('unhandled', 'done', 'archive') will resolve the matter.

Referring to this article
https://uxdesign.cc/best-ux-practices-for-search-inputs-c44dba565448
to distinguish search input from the basic input I have added magnifier icon on the left.
I have also added some more space under the search input. 

![End_of_Coffee_-_track_your_office_inventory](https://user-images.githubusercontent.com/27632432/65747901-ee8e5400-e102-11e9-8f76-94d600cc9898.png)

marekrozmus commented on client/src/assets/images/icons/index.js:
```diff
@@ -210,3 +210,12 @@ export const ArrowUpIcon = () => (
 );
+
+export const SearchIcon = () => (
+  <Icon data-icon="arrow-up" viewBox="0 0 512 512">
```
data-icon is incorrect. What is this prop for anyway?

adamklepacz replied:
it is here just for accessibility reasons. 
Change it to`dataIcon` or remove it anyway?

marekrozmus replied:
"arrow-up" should be "search"

marekrozmus commented on client/src/modules/list/components/ItemsContainer.scss:
```diff
@@ -146,4 +143,69 @@
     line-height: 1.2;
   }
+
+  &__filter {
```
why this style is here and not in `Filter.scss`?

adamklepacz replied:
I've been thinking about adding this style to a separated file but @alemikolo said that we may want to have different styles for this component in different places.

@alemikolo  can you refer to that?

alemikolo replied:
I thought that this component the component will only have the functionality implemented. Because we may want to style it in a different way in different places.

We use it to filter lists in `move the item to list`  panel, now its added to `itemsList` to filter items.  Łukasz said that he would like to have such kind of filter in activities panel.

In our desktop view in all those places, the filter should have at least slightly different styles...

Maybe it would be better to put filter styles in a separate file, and overwrite them only when it is necessary?

marekrozmus replied:
Move to component css file all default styles and override in other places elements that needs that.

Ciunkos commented on client/src/assets/images/icons/index.js:
```diff
@@ -210,3 +210,12 @@ export const ArrowUpIcon = () => (
 );
+
+export const SearchIcon = () => (
+  <Icon data-icon="search-icon" viewBox="0 0 512 512">
```
https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA24

adamklepacz replied:
Added as separated task

Ciunkos commented on client/src/common/components/Filter.jsx:
```diff
@@ -1,7 +1,10 @@
 import PropTypes from 'prop-types';
 import _debounce from 'lodash/debounce';
-import _trim from 'lodash/trim';
+import _trimStart from 'lodash/trimStart';
```
Please trim only before performing the search and always leave the input as it is entered by the user (make sure filter "list 2   " ["list 1", "list 2"] yields ["list 2"]).

Ciunkos commented on client/src/common/components/Filter.jsx:
```diff
@@ -22,14 +25,29 @@ class Filter extends PureComponent {
-    const { options, onFilter } = this.props;
+  filterByFields = query => {
+    const { filterBy, options } = this.props;
+    const match = new RegExp(query, 'i');
```
Please use case-invariant (toLowerCase()/toLocaleLowerCase) `String.includes` search as regexp constructor requires proper escaping (try "*" as a search query) and can lead to REDOS.

Ciunkos commented on client/src/common/components/Filter.jsx:
```diff
@@ -22,14 +25,29 @@ class Filter extends PureComponent {
+    return options.filter(option => {
+      let testString = '';
+
+      filterBy.forEach(field => {
```
Wouldn't fields name be better? Consider `filterBy` -> `fields` rename.

Ciunkos commented on client/src/modules/list/components/ItemsContainer.jsx:
```diff
@@ -1,94 +1,46 @@
-    return filterBy === FilterOptionType.MY_ITEMS
-      ? items.filter(item => item.authorId === id)
-      : items;
+    this.setState({ items });
```
Why not pure render filtering? Are there any performance issues present?

adamklepacz replied:
It was a customer's wish to replace the filter and sort components with search input

Ciunkos replied:
This is rather a technical question as this should not affect the UI. I think you should strive to minimize state and if possible, use pure filtering instead.

adamklepacz replied:
Sorry, but I don't understand what you mean by pure filtering?

marekrozmus commented on client/src/common/components/Filter.jsx:
```diff
@@ -45,28 +64,17 @@ class Filter extends PureComponent {
-    const { query } = this.state;
 
-    this.debouncedFilter(query);
+    this.setState(() => ({ query: value }), () => this.debouncedFilter(value));
```
Do you need function if not using previous state?
Isn't this enough? `this.setState({ query: value }, () => this.debouncedFilter(value));`

Ciunkos commented on client/src/common/components/Filter.scss:
```diff
@@ -0,0 +1,67 @@
+    cursor: pointer;
+
+    svg {
+      height: 14px;
```
Please use constant-size helper as svgs tends to collapse within flex layouts.

Ciunkos commented on client/src/common/components/Filter.scss:
```diff
@@ -0,0 +1,67 @@
+
+    &:hover {
+      svg {
+        height: 20px;
```
Please use `transform: scale` to use GPU-accelerated transitions.

Ciunkos commented on client/src/modules/list/components/Items/MoveToListPanel/MoveToListPanel.jsx:
```diff
@@ -118,6 +118,7 @@ class MoveToListPanel extends PureComponent {
       intl: { formatMessage },
       lists
     } = this.props;
+    const filterByFields = ['name'];
```
This can be moved outside the class context.

Ciunkos commented on client/src/modules/list/components/ItemsContainer.jsx:
```diff
@@ -112,11 +64,10 @@ class ItemsContainer extends Component {
-    const filteredList = this.filterItems(items, filterBy);
-    const sortedList = this.sortItems(filteredList, sortBy, sortOrder);
+    const { items } = this.state;
+    const filterByFields = ['authorName', 'name'];
```
This can be moved outside the class context.

marekrozmus commented on client/src/common/components/Filter.jsx:
```diff
@@ -22,14 +25,30 @@ class Filter extends PureComponent {
+    });
+  };
+
+  handleFilter = query => {
```
`handleFilterChange`? or what does this method do exactly

adamklepacz replied:
name changed
